### ISSUE TITLE
Hilbert SFC Partitioner test fails sporadically on OSX

### ISSUE THREAD

**Issue by jwpeterson** [libMesh/libmesh#1967]
These partitioner unit tests were originally merged in 0dc5c33ba (Oct 25) and for some reason they haven't failed on any of the CIVET testing we've done in the interim. I can get the Hilbert SFC test to fail fairly reliably on my Mac (but not every time, maybe some type of memory corruption?). It may be a platform-dependent issue. Unfortunately I can't seem to get a good stack trace down in the Hilbert library code itself-the last line of my trace, `hilbert.c:0`, is completely bogus. The error itself is a `SIGABRT`, so there's not much to go on.

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff66b5cb86 libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff66c12c50 libsystem_pthread.dylib`pthread_kill + 285
    frame #2: 0x00007fff66ac6268 libsystem_c.dylib`__abort + 144
    frame #3: 0x00007fff66ac6af6 libsystem_c.dylib`__stack_chk_fail + 205
    frame #4: 0x00000001031c3568 libmesh_dbg.0.dylib`hilbert(x=0x000000010a408bb0, y=0x000000010a408c90, z=0x000000010a408d70, N=0x00007ffeefbfb894, table=0x000000010a408e50) at hilbert.c:0
    frame #5: 0x000000010284d9f3 libmesh_dbg.0.dylib`libMesh::SFCPartitioner::partition_range(this=0x00007ffeefbfd998, mesh=0x00007ffeefbfda40, beg=element_iterator @ 0x00007ffeefbfcb28, end=element_iterator @ 0x00007ffeefbfcb08, n=2) at sfc_partitioner.C:125
    frame #6: 0x000000010284e96d libmesh_dbg.0.dylib`libMesh::SFCPartitioner::_do_partition(this=0x00007ffeefbfd998, mesh=0x00007ffeefbfda40, n=2) at sfc_partitioner.C:173
    frame #7: 0x00000001005730fe tmp-dbg`libMesh::HilbertSFCPartitioner::_do_partition(this=0x00007ffeefbfd998, mesh=0x00007ffeefbfda40, n=2) at hilbert_sfc_partitioner.h:75
    frame #8: 0x00000001027f3ac5 libmesh_dbg.0.dylib`libMesh::Partitioner::partition(this=0x00007ffeefbfd998, mesh=0x00007ffeefbfda40, n=2) at partitioner.C:86
    frame #9: 0x0000000100576e62 tmp-dbg`PartitionerTest<libMesh::HilbertSFCPartitioner, libMesh::ReplicatedMesh>::testPartition(this=0x000000010842f9a0, n_parts=2) at partitioner_test.h:57
    frame #10: 0x0000000100571f7a tmp-dbg`PartitionerTest<libMesh::HilbertSFCPartitioner, libMesh::ReplicatedMesh>::testPartition2(this=0x000000010842f9a0) at partitioner_test.h:101
    frame #11: 0x0000000100576406 tmp-dbg`CppUnit::TestCaller<PartitionerTest_HilbertSFCPartitioner_ReplicatedMesh>::runTest(this=0x000000010842f950) at TestCaller.h:166
    frame #12: 0x0000000106f4fa52 libcppunit-1.12.1.dylib`CppUnit::TestCaseMethodFunctor::operator()() const + 34
    frame #13: 0x0000000106f3eca3 libcppunit-1.12.1.dylib`CppUnit::DefaultProtector::protect(CppUnit::Functor const&, CppUnit::ProtectorContext const&) + 35
    frame #14: 0x0000000106f49e8c libcppunit-1.12.1.dylib`CppUnit::ProtectorChain::protect(CppUnit::Functor const&, CppUnit::ProtectorContext const&) + 364
    frame #15: 0x0000000106f58716 libcppunit-1.12.1.dylib`CppUnit::TestResult::protect(CppUnit::Functor const&, CppUnit::Test*, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 54
    frame #16: 0x0000000106f4f723 libcppunit-1.12.1.dylib`CppUnit::TestCase::run(CppUnit::TestResult*) + 227
    frame #17: 0x0000000106f4ff65 libcppunit-1.12.1.dylib`CppUnit::TestComposite::doRunChildTests(CppUnit::TestResult*) + 85
    frame #18: 0x0000000106f4fe4f libcppunit-1.12.1.dylib`CppUnit::TestComposite::run(CppUnit::TestResult*) + 31
    frame #19: 0x0000000106f4ff65 libcppunit-1.12.1.dylib`CppUnit::TestComposite::doRunChildTests(CppUnit::TestResult*) + 85
    frame #20: 0x0000000106f4fe4f libcppunit-1.12.1.dylib`CppUnit::TestComposite::run(CppUnit::TestResult*) + 31
    frame #21: 0x0000000106f584c2 libcppunit-1.12.1.dylib`CppUnit::TestResult::runTest(CppUnit::Test*) + 34
    frame #22: 0x0000000106f5c583 libcppunit-1.12.1.dylib`CppUnit::TestRunner::run(CppUnit::TestResult&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 51
    frame #23: 0x0000000106f5e2c9 libcppunit-1.12.1.dylib`CppUnit::TextTestRunner::run(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, bool, bool) + 73
    frame #24: 0x00000001000011f5 tmp-dbg`main(argc=4, argv=0x00007ffeefbfe380) at driver.C:24
    frame #25: 0x00007fff66a1e085 libdyld.dylib`start + 1
```
I'd like to temporarily skip this test until we have a chance to figure out what's going on. I'll put a reference to this PR in a comment so someone trying to re-enable it can see what's been found so far...

**Comment by jwpeterson** [comment]
I don't have a working Valgrind on my Mac, which unfortunately is the only place I can trigger the error...

**Comment by jwpeterson** [comment]
> Maybe kick the Mac test over and over again and see if any of them die?

That is possible but it would be pretty inefficient. The Mac target runs through all of libmesh make check and all of the MOOSE testing in both serial and parallel... if I can reproduce it on my laptop in addition to my workstation, I'll let you know?

**Comment by jwpeterson** [comment]
Three data points: 

1. The sfcurves partitioners are only enabled if you configure libmesh with `--disable-strict-lgpl`. I'm pretty sure we do have an optional recipe for this, but we don't run it very often on CIVET and this probably explains why we haven't seen the issue previously. In that case, it may not be a Mac v. Linux thing after all.

1. I found that if I put `printf` statements right up to the end of the `hilbert()` function in [hilbert.c](https://github.com/libMesh/libmesh/blob/master/contrib/sfcurves/hilbert.c) and then immediately following the `hilbert()` call back in [sfc_partitioner.C](https://github.com/libMesh/libmesh/blob/master/src/partitioning/sfc_partitioner.C), then I see all of the print statements from the C side but no print statement back on the C++ side, only the `SIGABRT`. This makes me think something terrible has happened to the stack during the call to `hilbert()`.

1. If I comment out the function call that does all the insane bit twiddling, i.e.
https://github.com/libMesh/libmesh/blob/7e30bf0a9ef28e7f8657864748fc5497dbac02c5/contrib/sfcurves/hilbert.c#L185
 I no longer get any aborts! Since that function is doing some very complicated array indexing/writing, I'd wager the problem is somewhere in there... no idea how to actually find it, though.

**Comment by jwpeterson** [comment]
On Linux, running the unit tests through valgrind, I get a segfault and the following "Invalid read" error:
```
Invalid read of size 4
   at 0x6B6CE39: hilbert (hilbert.c:193)
   by 0x65C9AEC: libMesh::SFCPartitioner::partition_range(libMesh::MeshBase&, libMesh::MeshBase::element_iterator, libMesh::MeshBase::element_iterator, unsigned int) (sfc_partitioner.C:125)
   by 0x65CA2BC: libMesh::SFCPartitioner::_do_partition(libMesh::MeshBase&, unsigned int) (sfc_partitioner.C:185)
   by 0x81C85D: libMesh::HilbertSFCPartitioner::_do_partition(libMesh::MeshBase&, unsigned int) (hilbert_sfc_partitioner.h:75)
   by 0x65A7574: libMesh::Partitioner::partition(libMesh::MeshBase&, unsigned int) (partitioner.C:86)
   by 0x81E8F6: PartitionerTest<libMesh::HilbertSFCPartitioner, libMesh::ReplicatedMesh>::testPartition(unsigned int) (partitioner_test.h:61)
   by 0x81C239: PartitionerTest<libMesh::HilbertSFCPartitioner, libMesh::ReplicatedMesh>::testPartition2() (partitioner_test.h:105)
   by 0x81E115: CppUnit::TestCaller<PartitionerTest_HilbertSFCPartitioner_ReplicatedMesh>::runTest() (TestCaller.h:166)
   by 0xEEFE31E: CppUnit::TestCaseMethodFunctor::operator()() const (TestCase.cpp:32)
   by 0xEEF50CF: CppUnit::DefaultProtector::protect(CppUnit::Functor const&, CppUnit::ProtectorContext const&) (DefaultProtector.cpp:15)
   by 0xEEFBCE3: CppUnit::ProtectorChain::protect(CppUnit::Functor const&, CppUnit::ProtectorContext const&) (ProtectorChain.cpp:77)
   by 0xEF04F76: CppUnit::TestResult::protect(CppUnit::Functor const&, CppUnit::Test*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (TestResult.cpp:178)
 Address 0xd8002b34 is not stack'd, malloc'd or (recently) free'd
```
The line of code in question is pretty innocuous looking, it's just accessing the i'th entry of the input `table`, and the i'th entry of `s`, neither of which should be invalid IMO... I'll need to keep on digging.
https://github.com/libMesh/libmesh/blob/7e30bf0a9ef28e7f8657864748fc5497dbac02c5/contrib/sfcurves/hilbert.c#L180-L195

**Comment by roystgnr** [comment]
```table``` is on the stack after the other function arguments... but none of those are addressed by index, only their contents are, so there's no obvious route for stack smashing there.  ```s``` is on the stack after ```index```, ```extrx``` etc...  ```extrx/extry/extrz``` are all accessed by constants within their array bounds...  Is it possible that something in ```fhsfc3d``` tries to write to ```index[16]```, corrupting ```s```?  No, wait, that would cause a failure on line 187, not 193.  Could we be passing in an insufficiently large buffer for ```table```?

**Comment by roystgnr** [comment]
Can you try -fstack-protector or whatever with clang and see if it screams earlier, in the return from the fhsfc3d call?

**Comment by jwpeterson** [comment]
> `s` is on the stack after `index`,

`s` is malloc'd, so it's on the heap...

> Could we be passing in an insufficiently large buffer for `table`?

Using print statements, I have determined that `table` has exactly 27 entries (one per elem centroid if I understand correctoy). Should it have more space than that? 
```
In hilbert, N=27
In hilbert, x=0.166667 0.500000 0.833333 0.166667 0.500000 0.833333 0.166667 0.500000 0.833333 0.166667 0.500000 0.833333 0.166667 0.500000 0.833333 0.166667 0.500000 0.833333 0.166667 0.500000 0.833333 0.166667 0.500000 0.833333 0.166667 0.500000 0.833333 
In hilbert, y=0.166667 0.166667 0.166667 0.500000 0.500000 0.500000 0.833333 0.833333 0.833333 0.166667 0.166667 0.166667 0.500000 0.500000 0.500000 0.833333 0.833333 0.833333 0.166667 0.166667 0.166667 0.500000 0.500000 0.500000 0.833333 0.833333 0.833333 
In hilbert, z=0.166667 0.166667 0.166667 0.166667 0.166667 0.166667 0.166667 0.166667 0.166667 0.500000 0.500000 0.500000 0.500000 0.500000 0.500000 0.500000 0.500000 0.500000 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 
In hilbert, table=0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
In hilbert, starting for loop from 1 to N.
In hilbert, starting for loop from 0 to N.
In hilbert, calling qsort.
In hilbert, final loop from 0 to N.
In hilbert, before free(s).
In hilbert, after free(s).
In hilbert, x=0.166667 0.166667 0.166667 0.166667 0.500000 0.500000 0.500000 0.500000 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.166667 0.166667 0.500000 0.500000 0.500000 0.166667 0.833333 0.833333 0.833333 0.500000 0.500000 0.166667 0.166667 
In hilbert, y=0.166667 0.166667 0.500000 0.500000 0.500000 0.500000 0.166667 0.166667 0.166667 0.166667 0.500000 0.500000 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.500000 0.166667 0.166667 0.500000 0.500000 0.166667 
In hilbert, z=0.166667 0.500000 0.500000 0.166667 0.166667 0.500000 0.500000 0.166667 0.166667 0.500000 0.500000 0.166667 0.166667 0.500000 0.500000 0.166667 0.166667 0.500000 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 0.833333 
In hilbert, table=1 10 13 4 5 14 11 2 3 12 15 6 9 18 16 7 8 17 26 25 27 24 21 20 23 22 19 
```

**Comment by jwpeterson** [comment]
I'm using GCC 7.3.0 so I'll try `-fstack-protector`; it appears that [clang also supports the same flag](https://stackoverflow.com/questions/39371514/stack-protect-equivalent-in-clang-compiler).

**Comment by jwpeterson** [comment]
No change (either segfault or invalid read error from valgrind) for me with `-fstack-protector` set in both `libmesh_CXXFLAGS` and `CFLAGS`. Not sure if that's because the compiler heuristically chooses which functions to stack protect (no buffer at least 8 bytes long, etc.) and this particular function in hilbert.c is simply not chosen? clang seems to have this heuristic as well...

**Comment by jwpeterson** [comment]
@rwcarlsen and I narrowed down the buffer overrun to line 72 in hilbert.c:
https://github.com/libMesh/libmesh/blob/7e30bf0a9ef28e7f8657864748fc5497dbac02c5/contrib/sfcurves/hilbert.c#L68-L73
the index `which` is actually 3 in some cases when we enter the `else` clause, but `key` only contains 3 entries. We can "fix" the problem by changing the if-statement to something simpler, e.g.
```
if (which > 2) {
```
and this seems to give, at least for this unit test, the same answers the test previously gave before (when we got lucky and it actually ran instead of segfaulting). I'd like to take a closer look at the partitions the modified code produces for different problems, and if they still seem reasonable then I'll make a PR with the proposed fix...
